### ISSUE TITLE
Zig Zag indicator implementation question

### ISSUE THREAD

**Issue by francescopeloi** [ta4j/ta4j#1122]
Hello, I am not yet an expert of indicators and neither ta4j, and trying to develop my first indicator: the Zig Zag indicator.

This indicator seems to me peculiar, since with new close price values coming in, it might change previous calculated indicator values. This is due to the logic this indicator works.

In a nutshell:
```
data = {1.0, 1.1, 1.4}
zigzag(0) = 1.0
zigzag(2) = 1.4
```
in this case there's a (zigzag) line from 1.0 to 1.4 
```
data = {1.0, 1.1, 1.4, 2.0}
zigzag(0) = 1.0
zigzag(3) = 2.0
```
in this case, where a new value of 2.0 comes in, the (zigzag) line is drawn from 1.0 to 2.0, which implies that also previous values are different, as the line inclination is now different. So `zigzag(2)` is **not** going to be 1.4 anymore, like stated in the previous example.

Give the above, in a test like this: 
```
BarSeries data = new MockBarSeries(numFunction, 1.0, 1.0, 1.3, 1.5, 1.6, 1.8 , 0.9, 0.8, 0.8, 0.7, 1.0, 1.1, 1.1, 1.2);
```
where I provide all the test data in advance, should the indicator know all values? Or it should answer knowing the values only "incrementally"?

In other words: `zigzag(4)` knows only data[] from `data[0]` to `data[4]` or it knows data from `data[0]` to `data[data.length]`?

Also do you know any already implemented indicator that works similarly so I can take inspiration?

I have other questions but will start with these and see if I get traction :smile: 

Thanks for your help.

**Comment by francescopeloi** [comment]
I think I answered myself.

> where I provide all the test data in advance, should the indicator know all values? Or it should answer knowing the values only "incrementally"?

the indicator should know all the values. The barSerie the indicator provides has a beginIndex and endIndex. Even if the caller is asking for e.g. `zigzag(1)`, endIndex could be `10`, which means that further points are provided and should be taken into account in the indicator calculations.

> Also do you know any already implemented indicator that works similarly so I can take inspiration?
I've found [DifferencePercentageIndicator](https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/DifferencePercentageIndicator.java) that seems similar/useful. I don't get if the use of instance variables [lastNotification](https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/DifferencePercentageIndicator.java#L46), but it seems that it has been left there after the [refactoring](https://github.com/ta4j/ta4j/pull/1060) to remove instance variables. Still have to figure out why.

I'll continue developing with these assumptions.

**Comment by TheCookieLab** [comment]
@francescopeloi you are correct, the indicator should be aware of all the values. There actually used to be a ZigZagIndicator class in the project but it was removed at some point. I don't know the reason but I assume it's because it didn't work quite as expected. That said here is the original code which may provide some value for you (i.e. things to borrow, things to avoid, etc) during your implementation:

```

package org.ta4j.core.indicators;

import org.ta4j.core.Indicator;
import org.ta4j.core.num.NaN;
import org.ta4j.core.num.Num;

/**
 * The Zig Zag Indicator.
 * <p>
 * For more information, see
 * <a href="https://www.investopedia.com/terms/z/zig_zag_indicator.asp">Zig Zag
 * Indicator</a>.
 */
public class ZigZagIndicator extends CachedIndicator<Num> {

    /**
     * The threshold ratio (positive number).
     */
    private final Num thresholdRatio;

    /**
     * The indicator to provide values.
     * <p>
     * Can be {@link org.ta4j.core.indicators.helpers.ClosePriceIndicator} or
     * something similar.
     */
    private final Indicator<Num> indicator;

    private Num lastExtreme;

    /**
     * Constructs a ZigZagIndicator.
     *
     * @param indicator the indicator to provide values
     * @param thresholdRatio the threshold ratio, must be positive
     */
    public ZigZagIndicator(Indicator<Num> indicator, Num thresholdRatio) {
        super(indicator);
        this.indicator = indicator;
        validateThreshold(thresholdRatio);
        this.thresholdRatio = thresholdRatio;
    }

    /**
     * Validates the provided threshold.
     *
     * @param threshold the threshold to validate
     * @throws IllegalArgumentException if threshold is null or non-positive
     */
    private void validateThreshold(Num threshold) {
        if (threshold == null) {
            throw new IllegalArgumentException("Threshold ratio is mandatory");
        }
        if (threshold.isNegativeOrZero()) {
            throw new IllegalArgumentException("Threshold ratio value must be positive");
        }
    }

    /**
     * Calculates the Zig Zag indicator's value for the current index.
     *
     * @param index the bar index
     * @return the indicator's value if its ratio from the last extreme is equal
     * to or greater than the threshold, otherwise {@link NaN}
     */
    @Override
    protected Num calculate(int index) {
        if (index == 0) {
            lastExtreme = indicator.getValue(0);
            return lastExtreme;
        } else {
            if (lastExtreme.isZero()) {
                // Treat this case separately
                // because one cannot divide by zero
                return NaN.NaN;
            } else {
                Num indicatorValue = indicator.getValue(index);
                Num differenceRatio = indicatorValue.minus(lastExtreme).dividedBy(lastExtreme);

                if (differenceRatio.abs().isGreaterThanOrEqual(thresholdRatio)) {
                    lastExtreme = indicatorValue;
                    return lastExtreme;
                } else {
                    return NaN.NaN;
                }
            }
        }
    }
}



```

**Comment by francescopeloi** [comment]
thanks for your answer @TheCookieLab . I found that indicator you posted, but couldn't make it work properly, so started from scratch. I think this indicator can be implemented easily in a naive way, the problem is that if you don't look close, you're not going to find some hard-to-find edge cases that if you don't solve, what you get is just wrong (some examples: 2 high zigzag points after each other, you need to re-draw the previous line; or another one: price is exactly the same for some time: is it a high point o a low swing point?). I am in the middle of trying to solve them in my first implementation, than I have to step back and see the whole picture again and probably write it again, as the code now looks awful :) 

Can you help me understand one thing: I've found [DifferencePercentageIndicator](https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/DifferencePercentageIndicator.java) that seems similar. I want to undestand how it works to get some inspiration. I don't get if the use of the instance variable [lastNotification](https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/DifferencePercentageIndicator.java#L46) there is legit, given that there was a plan to https://github.com/ta4j/ta4j/issues/906, but it seems that it has been left there after the https://github.com/ta4j/ta4j/pull/1060 to remove instance variables.

Do you know why that instance variable would be acceptable? I thought that all instance variables should be removed from all indicators, given that even if you are using one, you're breaking statelessness, which we should not.

**Comment by luizlucca** [comment]
@francescopeloi Can you correctly implement this indicator?
If so, could you share the code, please?

**Comment by TheCookieLab** [comment]
> thanks for your answer @TheCookieLab . I found that indicator you posted, but couldn't make it work properly, so started from scratch. I think this indicator can be implemented easily in a naive way, the problem is that if you don't look close, you're not going to find some hard-to-find edge cases that if you don't solve, what you get is just wrong (some examples: 2 high zigzag points after each other, you need to re-draw the previous line; or another one: price is exactly the same for some time: is it a high swing point o a low swing point? and which of those?). I am in the middle of trying to solve them in my first implementation, than I have to step back and see the whole picture again and probably write it again, as the code now looks awful :)
> 
> Can you help me understand one thing: I've found [DifferencePercentageIndicator](https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/DifferencePercentageIndicator.java) that seems similar. I want to undestand how it works to get some inspiration. I don't get if the use of the instance variable [lastNotification](https://github.com/ta4j/ta4j/blob/master/ta4j-core/src/main/java/org/ta4j/core/indicators/helpers/DifferencePercentageIndicator.java#L46) there is legit, given that there was a plan to remove them (#906), but it seems that it has been left there after the PR (#1060) to remove instance variables.
> 
> Do you know why that instance variable would be acceptable? I thought that all instance variables should be removed from all indicators, given that even if you are using one, you're breaking statelessness, which we should not.

You are right, the `lastNotification` variable should be removed. I don't know the history of this particular indicator but my guess is `lastNotification` was either missed or removing it was hairier than expected. This is something that will need resolution.

**Comment by francescopeloi** [comment]
hi @luizlucca and @TheCookieLab ,

I haven't found a way to make the ZigZag indicator work in ta4j. I can plot it fine, but cannot make it work in ta4j strategies. I'll try to explain why:

For its nature, the last point of the ZigZag indicator is always the last price of the price that is used (e.g. close price). All zigzag indicator points before are recalculated based on this new last point. In fact, if you watch a zigzag indicator live, you can see that the last segment is constantly changing, and sometimes (when some conditions are met) also previous zigzag points change. This is how the zigzag indicator works.

Now in the ta4j strategy framework, the strategy can ask for its indicators value at any point. This is I believe how ta4j is designed. E.g.: your strategy is checking what to do at barseries index 70, the strategy rules might check its internal indicators at index 69 but also 64, in no particular order, 64 might be called before 69. This is problematic for the zigzag, because you should first fix the zigzag at the last index (70) and then get the values from this instance.

Not sure if that's clear, but I don't know how to get around this problem, any suggestion welcome.

**Comment by TheCookieLab** [comment]
Thanks for bringing this up. I don't have any ideas off the top of my head
but will do a deep dive in order to better grok the situation.

**Comment by francescopeloi** [comment]
Basically the ZigZag indicator wants to know 2 indexes:

- the last index of the bar series, so it knows all the data it's got
- the index of the bar series that it needs to return the value for

so I'd imagine a new method in the indicator:
```
T getValue(int index, int endIndex);
```

Currently my dirty workaround is:
```
public class PrecalculatedIndicatorStrategy extends BaseStrategy {

    private final PrecalculatedIndicator<Num> precalculatedIndicator;

    public PrecalculatedIndicatorStrategy(String name, Rule entryRule, Rule exitRule, PrecalculatedIndicator<Num> precalculatedIndicator, int unstableBars) {
        super(name, entryRule, exitRule, unstableBars);
        this.precalculatedIndicator = precalculatedIndicator;
    }

    @Override
    public boolean shouldEnter(int index, TradingRecord tradingRecord) {
        precalculateIndicator(index);
        return super.shouldEnter(index, tradingRecord);
    }

    @Override
    public boolean shouldExit(int index, TradingRecord tradingRecord) {
        precalculateIndicator(index);
        return super.shouldExit(index, tradingRecord);
    }

    private void precalculateIndicator(int index) {
        var indicatorSimpleName = getPrecalculatedIndicator().getClass().getSimpleName();
        log.trace("Precalculating indicator {} at index {}", indicatorSimpleName, index);
        precalculatedIndicator.precalculate(index);
    }

}
```

and 
```
public interface PrecalculatedIndicator<T> extends Indicator<T> {

    void precalculate(int index);
}
```

`precalculates` prepares the data, then the usual `getValue` returns the right index value.

But this breaks statelessness and of course is a dirty hack.

**Comment by TheCookieLab** [comment]
OK so I've kept my comments thus far at a high/general level largely due to never having used the ZigZag indicator. A point of clarification you mentioned:

> For its nature, the last point of the ZigZag indicator is always the last price of the indicator that is used (e.g. close price)

Do this mean that calling `ZigZagIndicator.getValue(barSeriesEndIndex)` would simply return `barSeries.getLastBar().getClosePrice()`?

I found a couple implementations of ZigZag on TradingView and this one stood out:

![BTCUSD_2024-08-30_09-21-01_c920e](https://github.com/user-attachments/assets/d45d659a-a262-4184-9707-3be1d2fa5fd4)

Here it looks like the last ZigZag value is actually the previously found high or low. Would this approach make sense for your use case and if so, would it be easier to implement?

**Comment by francescopeloi** [comment]
mmm well it depends. This ZigZag is not calculating the last few values (from the last high zigzag to "now"). The ZigZag I've seen (InteractiveBrokers) do calculate it until the last values. If it's meaningful or not, I'd redirect this question to someone else :D but my guess is that you'd also be interested in the most recent values, not only the values until the last high/low zigzag. But again it's just my guess, maybe a question for a more experienced trader.

**Comment by francescopeloi** [comment]
note that in your example, the last found high zigzag might change, if a newer higher than this high zigzag price comes, that will be the new high zigzag. This is what makes this indicator "difficult": previous values might change.

**Comment by TheCookieLab** [comment]
Gotcha, so how far back does your precalculatedIndicator logic go, just past the last zig zag or further?

**Comment by francescopeloi** [comment]
At the moment I re-draw everything all the time, which is obviously inefficient. This is a reasoning I still need to do. If you calculated already some previous zigzag points, I think you might need to-redraw just from the last few (todo: to determine how many, maybe 2?), because the others _I think_ will not change. I have a task on my list to figure this out, but didn't get to that yet.

**Comment by francescopeloi** [comment]
At the moment I am not implementing `CachedIndicator`, not sure if I written that somewhere? My `PrecalculatedIndicator` extends `Indicator`.

**Comment by malakeel** [comment]
@francescopeloi 

Any updates on sharing this work ?

**Comment by francescopeloi** [comment]
hi @malakeel , if you read previous messages you'll find out that currently this cannot be cleanly implemented in ta4j.

**Comment by TheCookieLab** [comment]
Closing as not planned for ta4j core. A repainting ZigZag model conflicts with ta4j’s deterministic index-based strategy evaluation semantics unless we introduce substantial framework changes. If needed, this is better maintained as an external/custom indicator implementation.
